aiohttp/connector.py

```diff
@@ -676,37 +669,32 @@ def clear_dns_cache(self, host=None, port=None):
         if (key in self._cached_hosts) and \
                 (not self._cached_hosts.expired(key)):
 
-            if traces:
-                for trace in traces:
-                    await trace.send_dns_cache_hit(host)
+            if lifecycle:
+                await lifecycle.send_dns_cache_hit(host)
 
             return self._cached_hosts.next_addrs(key)
 
         if key in self._throttle_dns_events:
-            if traces:
-                for trace in traces:
-                    await trace.send_dns_cache_hit(host)
+            if lifecycle:
+                await lifecycle.send_dns_cache_hit(host)
             await self._throttle_dns_events[key].wait()
         else:
-            if traces:
-                for trace in traces:
-                    await trace.send_dns_cache_miss(host)
+            if lifecycle:
+                await lifecycle.send_dns_cache_miss(host)
             self._throttle_dns_events[key] = \
                 EventResultOrError(self._loop)
             try:
 
-                if traces:
-                    for trace in traces:
-                        await trace.send_dns_resolvehost_start(host)
+                if lifecycle:
+                    await lifecycle.send_dns_resolvehost_start(host)
```

Comment:
I think it would be better to have a context class for stuff like this that's a no-op if lifecycle is None, otherwise do the right thing at start and end, would make things cleaner too.

Reply:
Agree, we could either make a specific `with lifecycle.send_dns_resolvehost_context():` for each pair of start/end signals or even use a generic `with lifecycle. enter_context(start="send_dns_resolvehost_start", end="send_dns_resolvehost_end", exception="send_dns_resolvehost_exception")`.